src/Dictionaries/RegExpTreeDictionary.cpp

```diff
@@ -0,0 +1,545 @@
+            if (structure.hasAttribute(name))
+            {
+                const auto & attr = structure.getAttribute(name);
+                auto string_pieces = createStringPieces(value, num_captures, regex, logger);
+                if (!string_pieces.empty())
+                {
+                    node->attributes[name] = RegexTreeNode::AttributeValue{.field = values[i], .pieces = std::move(string_pieces)};
+                }
+                else
+                {
+                    Field field = parseStringToField(values[i].safeGet<String>(), attr.type);
+                    node->attributes[name] = RegexTreeNode::AttributeValue{.field = std::move(field)};
+                }
+            }
+        }
+        regex_nodes.emplace(id, std::move(node));
+    }
+}
+
+void RegExpTreeDictionary::initGraph()
+{
+    for (const auto & [id, value]: regex_nodes)
+    {
+        UInt64 pid = value->parent_id;
+        if (pid == 0) // this is root
+            continue;
+        if (regex_nodes.contains(pid))
+            regex_nodes[pid]->children.push_back(id);
+        else
+            throw Exception(ErrorCodes::INCORRECT_DICTIONARY_DEFINITION, "Unknown parent id {}", pid);
+    }
+    std::set<UInt64> visited;
+    UInt64 topology_id = 0;
+    for (const auto & [id, value]: regex_nodes)
+        if (value->parent_id == 0) // this is root node.
+            initTopologyOrder(id, visited, topology_id);
+    if (topology_order.size() != regex_nodes.size())
+        throw Exception(ErrorCodes::INCORRECT_DICTIONARY_DEFINITION, "Invalid Regex tree");
+}
+
+void RegExpTreeDictionary::initTopologyOrder(UInt64 node_idx, std::set<UInt64> & visited, UInt64 & topology_id)
+{
+    visited.insert(node_idx);
+    for (UInt64 child_idx : regex_nodes[node_idx]->children)
+        if (visited.contains(child_idx))
+            throw Exception(ErrorCodes::INCORRECT_DICTIONARY_DEFINITION, "Invalid Regex tree");
+        else
+            initTopologyOrder(child_idx, visited, topology_id);
+    topology_order[node_idx] = topology_id++;
+}
+
+void RegExpTreeDictionary::loadData()
+{
+    if (!source_ptr->hasUpdateField())
+    {
+        QueryPipeline pipeline(source_ptr->loadAll());
+        PullingPipelineExecutor executor(pipeline);
+
+        Block block;
+        while (executor.pull(block))
+        {
+            initRegexNodes(block);
+        }
+        initGraph();
+    }
+    else
+    {
+        throw Exception(ErrorCodes::UNSUPPORTED_METHOD, "Dictionary {} does not support updating manual fields", name);
+    }
+}
+
+RegExpTreeDictionary::RegExpTreeDictionary(
+    const StorageID & id_, const DictionaryStructure & structure_, DictionarySourcePtr source_ptr_, Configuration configuration_)
+    : IDictionary(id_), structure(structure_), source_ptr(source_ptr_), configuration(configuration_), logger(&Poco::Logger::get("RegExpTreeDictionary"))
+{
+    if (auto * ch_source = typeid_cast<ClickHouseDictionarySource *>(source_ptr.get()))
```

Comment:
How sample_block will work if source is other than `Clickhouse` ? Is it safe to change `ClickHouseDictionarySource::sample_block` after `ClickHouseDictionarySource` was created?

Reply:
if source is not `ClickHouse`, the schema will mismatch and raise a Exception.

This is only for test. We assume that the user only uses YAML source by default.